### ISSUE TITLE
Co2L: Contrastive Continual Learning

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#136]
## Motivation

- *Forgetting* in continual learning framework occurs not only of accidents, but also because the representation learned from previous tasks are not useful for the new/current task. (Apple vs Banana / Apple vs Banana vs Strawberry)
- Contrastive learning has been shown to help learn more *robust and transferrable* knowledge, hence employ the concept in CL.

## TL; DR

- Propose a novel ***Rehearsal-based Continual Learning Framework*** that employs contrastive learning and self-distillation.
- Asymmetric Supervised Contrastive Loss: pull features from same task / class together, push others apart
- Instance-wise Relation Distillation: enforce relative sample similarities obtained from current model and previous model (from the last task) to be matched

## Previous Work

- Rehearsal-based Continual Learning Methods
- Contrastive Learning Methods (both supervised and unsupervised)

## Method

![image](https://user-images.githubusercontent.com/46308407/145926463-fecda588-078a-4663-a6e5-18373b4bcd84.png)
![image](https://user-images.githubusercontent.com/46308407/145926479-bbf78df6-c26e-41dc-87cf-4afec8855a1a.png)

### Assymetric Supervised Contrastive Learning

![image](https://user-images.githubusercontent.com/46308407/145926763-868a0f76-08e0-4e25-8d90-67eb63fea37a.png)

- Make two random augmentations as commonly done in SSL literature
- Take samples from the current task as anchors, and those that belong to the same task/class are positive data points
- Apply traditional InfoNCE loss to the samples, pulling those that belong together close together and pushing others far apart.

### Instance-wise Relation Distillation

![image](https://user-images.githubusercontent.com/46308407/145926800-ba7b0979-3a59-47af-b0d0-b70d5eb49842.png)
![image](https://user-images.githubusercontent.com/46308407/145926813-5b1f9ac0-874b-4af7-950b-4a7e10c9211f.png)

- To maintain (not forget) useful representations learned from the previous model (i.e., the model checkpoint from the final iteration of last task), enforce the relative sample similarities obtained from the features to be similar (similar idea to CDC)
- This is a means of self-distillation, in which the knowledge from previous model state can be effectively transferred to the current task model.
- The loss function takes the form of Cross Entropy, which is essentially equivalent to KL divergence, as formulated in CDC.

![image](https://user-images.githubusercontent.com/46308407/145926851-50508750-d059-4569-ad57-315b2f2c4bb2.png)

### Dataset

- **Seq-CIFAR-10**: CIFAR-10 split into 5 pairs\
- **Seq-Tiny-ImageNet**: 200 classes split into 10 20-class tasks
- **R-MNIST**: MNIST with random rotations (20 rotation degrees - 20 subtasks)

### Architecture

- ResNet-18 as backbone + Linear classifier on top of the frozen representation

### Transferrability of Contrastive Features

![image](https://user-images.githubusercontent.com/46308407/145926900-2645397e-dd72-4cee-bc13-8937faa7f500.png)

### Comparison with Baselines

![image](https://user-images.githubusercontent.com/46308407/145926943-886bee9c-6859-4391-95e6-23c86df56030.png)

### Ablation

![image](https://user-images.githubusercontent.com/46308407/145926966-81567546-2ebe-44a9-a4ed-49ca8d257d4e.png)
![image](https://user-images.githubusercontent.com/46308407/145926976-76e04718-b608-446c-bc84-29237e42739c.png)
